### PR TITLE
restrict node start-up when cluster name in data path

### PR DESCRIPTION
**Pull request by talevy**
When a 2.x cluster is created, the structure of `path.data` has all contents for a node inside a directory named after the cluster name. This was changed in 5.x (https://github.com/elastic/elasticsearch/pull/18554) to remove the directory with the cluster name and move its contents up a level. A 5.x cluster will still read the 2.x structure correctly. In 6.x this backwards compatible behavior is removed (https://github.com/elastic/elasticsearch/pull/20433), and if a 6.x node is started with a data directory using the old 2.x structure, it will see it as if it was empty and ignore the existing data.

This PR makes it so that a 6.x node refuses to start when there exists a data path 
with the cluster name in it.

relates: https://github.com/elastic/elasticsearch/issues/32661#issuecomment-432306352

### CODE REVIEW

**Comment by talevy** [comment]
cc/ @bleskes took me a bit of time to get to this only to realize what the tests show, we don't necessarily have permissions to check whether there is a cluster-name in the data path.

**Comment by bleskes** [comment]
Thx @talevy . I'm inclined to implement this as a `BootstrapCheck` that's always enforced. That said, those are run after the security manager is installed as well. @danielmitterdorfer can you please advise?

**Comment by danielmitterdorfer** [comment]
A boostrap check makes sense to me (although they will only issue a warning when bound to loopback but I think it is ok?). I also think that we are safe with a bootstrap check because we setup the necessary permissions to access the data paths in `Security.configure`and then invoke the bootstrap checks. As the path in question is in a subdirectory of the data path (to which we have full access) I think we should be fine w.r.t. to access checks?

**Comment by talevy** [comment]
thanks @danielmitterdorfer @bleskes, I will see how to make this a BootstrapCheck

**Comment by talevy** [comment]
Update: I've learned more about BootstrapChecks and realized there are a few nice refactors to do that will make writing the `ClusterNameInDataPathCheck` a lot cleaner. Long story short: This check needs access to the node's Environment to work right.

PR that needs to be merged before continuing: https://github.com/elastic/elasticsearch/pull/36573
related PR that would be a nice to have: https://github.com/elastic/elasticsearch/pull/36574

**Comment by talevy** [comment]
Update: refactors to give access to Environment#pathFiles have made it in, so now this is ready!

File: server/src/main/java/org/elasticsearch/bootstrap/BootstrapChecks.java [hunk 1]
```diff
@@ -712,4 +716,26 @@ boolean isAllPermissionGranted() {
         }
 
     }
+
+    /**
+     * Bootstrap check for cluster name directories in data paths
+     */
+    static class ClusterNameInDataPathCheck implements BootstrapCheck {
```
Comment (danielmitterdorfer) [inline server/src/main/java/org/elasticsearch/bootstrap/BootstrapChecks.java]: Should this overrride `alwaysEnforce()` and return `true`?

**Review by danielmitterdorfer — COMMENTED** [review]
Looks fine overall. I left a couple of suggestions / questions about the error message.

File: server/src/main/java/org/elasticsearch/bootstrap/BootstrapChecks.java [hunk 2]
```diff
@@ -712,4 +716,31 @@ boolean isAllPermissionGranted() {
...
+            final ClusterName clusterName = ClusterName.CLUSTER_NAME_SETTING.get(checkEnvironment.settings());
+            List<String> existingPathsWithClusterName = Arrays.stream(checkEnvironment.dataFiles())
+                .map(p -> p.resolve(clusterName.value()))
+                .filter(Files::exists).map(Path::toString).collect(Collectors.toList());
+            if (existingPathsWithClusterName.isEmpty()) {
+                return BootstrapCheckResult.success();
+            } else {
+                final String paths = String.join(",", existingPathsWithClusterName);
+                return BootstrapCheckResult.failure(
+                    "node cannot have cluster names as subdirectories in data paths [" + paths + "]");
```
[the remark this answers is not in the document]
Comment (talevy) [reply server/src/main/java/org/elasticsearch/bootstrap/BootstrapChecks.java]: thanks for the suggestion. more detail is probably best. updated

**Comment by talevy** [comment]
example run from a test cluster named `elasticsearch`

```
[2018-12-17T10:00:23,572][INFO ][o.e.n.Node               ] [1Xo7d5f] initialized
[2018-12-17T10:00:23,573][INFO ][o.e.n.Node               ] [1Xo7d5f] starting ...
[2018-12-17T10:00:23,796][INFO ][o.e.t.TransportService   ] [1Xo7d5f] publish_address {127.0.0.1:9300}, bound_addresses {[::1]:9300}, {127.0.0.1:9300}

ERROR: [1] bootstrap checks failed

[1]: Cluster name [elasticsearch] subdirectory exists in data paths [/distribution/archives/tar/build/distributions/elasticsearch-6.6.0-SNAPSHOT/data/elasticsearch]. All data under these paths must be moved up one directory to paths [/distribution/archives/tar/build/distributions/elasticsearch-6.6.0-SNAPSHOT/data]

[2018-12-17T10:00:23,850][INFO ][o.e.n.Node               ] [1Xo7d5f] stopping ...
[2018-12-17T10:00:23,865][INFO ][o.e.n.Node               ] [1Xo7d5f] stopped
[2018-12-17T10:00:23,866][INFO ][o.e.n.Node               ] [1Xo7d5f] closing ...
[2018-12-17T10:00:23,924][INFO ][o.e.n.Node               ] [1Xo7d5f] closed
[2018-12-17T10:00:23,928][INFO ][o.e.x.m.p.NativeController] [1Xo7d5f] Native controller process has stopped - no new native processes can be started
```

**Comment by talevy** [comment]
run the default distro tests

**Review by jasontedor — CHANGES_REQUESTED** [review]
The production code looks good, yet I would structure it a little differently?

File: server/src/main/java/org/elasticsearch/node/Node.java [hunk 3]
```diff
@@ -739,6 +739,20 @@ public Node start() throws NodeValidationException {
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
+
+        final List<Path> existingPathsWithClusterName = Arrays.stream(environment.dataFiles())
```
[the remark this answers is not in the document]
Comment (talevy) [reply server/src/main/java/org/elasticsearch/node/Node.java]: for sure, I was 50/50 on doing this, but decided against it due to fear of adding too much overhead. makes sense though, that method is large enough. I'll update

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by talevy** [comment]
thanks Jason!

**Comment by talevy** [comment]
and thanks @danielmitterdorfer for initial review and suggestions!

**Comment by ywelsch** [comment]
This PR was missing version and type labels. I've added them based on the commit that was merged. Please adapt if necessary.

**Pull request merged**